### ISSUE TITLE
pred_contribs=True handling of base_margin 

### ISSUE THREAD

**Issue by dan11mcguire** [dmlc/xgboost#6546]
For models with binary:logistic or reg:linear objective functions, 
the last column of the matrix obtained by
`model.predict(data, pred_contribs=True) `
, is equal to the mean of the predictions in training data.

But if I `set_base_margin()` on my training data.  For a `'count:poisson'` objective, for example, then my last column is equal to 
`data.get_base_margin() + 'bias'`

However, the 'bias' is not equal to the mean of predictions in training, i.e.

`bias != np.mean(model.predict(data,output_margin=True) - data.get_base_margin())`

Why is this, and how is base_margin handled together with pred_contribs=True?

**Comment by trivialfis** [comment]
@RAMitchell

**Comment by dan11mcguire** [comment]
Here's a simple numerical example to reproduce what I mean, if that helps.
The bias term does not seem to be correct when "set_base_margin" is used, this leads to incorrect pred_contribs (even on the same training data).  
Example ipynb simulates from a linear model with 3 independent covariates, shows the shap values are biased due to incorrect 'bias'/intercept term.  

[basemargin_treeshap_bias_show.zip](https://github.com/dmlc/xgboost/files/5983867/basemargin_treeshap_bias_show.zip)
![image](https://user-images.githubusercontent.com/49620432/107989192-bdd0e800-6f9f-11eb-8369-8f8e9905e306.png)

**Comment by nshahpazov-fourth** [comment]
Is there progress on this issue? I'm also someone to whom it would be of great help if this is included in the functionality of the `predict` method.

**Comment by NIkhilgKrish** [comment]
@trivialfis mind if i work on this issue. It sounds interesting I would like to that a crack at it.

**Comment by trivialfis** [comment]
Absolutely! Feel free to dig into the code, I confess I have difficulty of understanding the original SHAP dynamic programming algorithm from the paper.
